### PR TITLE
Fixed the problem of abnormal socket status during DNS resolution

### LINKED ISSUE #2848: Jedis connect failed when using dns resolver in jedis 4.1.0
### Expected behavior

Write here what you're expecting ...

connect success

### Actual behavior

Write here what happens instead ...

```
Exception in thread "main" redis.clients.jedis.exceptions.JedisConnectionException: Failed to connect to any host resolved for DNS name.
	at redis.clients.jedis.DefaultJedisSocketFactory.connectToFirstSuccsefulHost(DefaultJedisSocketFactory.java:62)
	at redis.clients.jedis.DefaultJedisSocketFactory.createSocket(DefaultJedisSocketFactory.java:85)
	at redis.clients.jedis.Connection.connect(Connection.java:180)
	at redis.clients.jedis.Connection.sendCommand(Connection.java:152)
	at redis.clients.jedis.Connection.sendCommand(Connection.java:135)
	at redis.clients.jedis.Jedis.ping(Jedis.java:353)
	at cn.nextop.gadget.redis.Main.main(Main.java:29)
	Suppressed: java.net.SocketTimeoutException: connect timed out
		at java.base/java.net.PlainSocketImpl.waitForConnect(Native Method)
		at java.base/java.net.PlainSocketImpl.socketConnect(PlainSocketImpl.java:107)
		at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:399)
		at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:242)
		at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:224)
		at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:403)
		at java.base/java.net.Socket.connect(Socket.java:591)
		at redis.clients.jedis.DefaultJedisSocketFactory.connectToFirstSuccsefulHost(DefaultJedisSocketFactory.java:65)
		... 6 more
	Suppressed: java.net.SocketException: Socket closed
		at java.base/java.net.PlainSocketImpl.checkAndReturnNativeFD(PlainSocketImpl.java:307)
		at java.base/java.net.PlainSocketImpl.socketConnect(PlainSocketImpl.java:91)
		at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:399)
		at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:242)
		at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:224)
		at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:403)
		at java.base/java.net.Socket.connect(Socket.java:591)
		at redis.clients.jedis.DefaultJedisSocketFactory.connectToFirstSuccsefulHost(DefaultJedisSocketFactory.java:65)
		... 6 more
```
### Steps to reproduce:

Please create a reproducible case of your problem. Make sure
that case repeats consistently and it's not random
```
try(Jedis jedis = new Jedis("dev-erebor.nextop.cn", 6379)) {
	System.out.println(jedis.ping()); // failed
}
```

dns resolver
```
InetAddress[] hosts = InetAddress.getAllByName("dev-erebor.nextop.cn");
for (InetAddress address : hosts) {
	System.out.println(address);
}

console print
dev-erebor.nextop.cn/240e:60c:2801:1:0:0:c0a8:1e9
dev-erebor.nextop.cn/192.168.1.233
```

when connect to `240e:60c:2801:1:0:0:c0a8:1e9` failed then connect to `192.168.1.233` also failed
but use IP address directly then connect success.
```
try(Jedis jedis = new Jedis("192.168.1.233", 6379)) {
	System.out.println(jedis.ping()); // success
}
```
### Redis / Jedis Configuration

redis.conf
```
bind 192.168.1.233
```

### PR DESCRIPTION
**Pull request by yangbodong22011**
fix #2848

### CODE REVIEW

File: src/main/java/redis/clients/jedis/DefaultJedisSocketFactory.java [hunk 1]
```diff
@@ -54,16 +54,23 @@ public DefaultJedisSocketFactory(HostAndPort hostAndPort, JedisClientConfig conf
...
     }
+
     JedisConnectionException jce = new JedisConnectionException("Failed to connect to any host resolved for DNS name.");
     for (InetAddress host : hosts) {
+      Socket socket = new Socket();
+      socket.setReuseAddress(true);
+      socket.setKeepAlive(true); // Will monitor the TCP connection is valid
+      socket.setTcpNoDelay(true); // Socket buffer Whetherclosed, to ensure timely delivery of data
+      socket.setSoLinger(true, 0); // Control calls close () method, the underlying socket is closed immediately
+      socket.setSoTimeout(socketTimeout);
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/DefaultJedisSocketFactory.java:70]: I understand this won't make much difference but IMO we should move this block within following `try` which would then be more similar to previous version of code.

**Review by dengliming — DISMISSED** [review]

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/redis/clients/jedis/DefaultJedisSocketFactory.java (hunk 1 above)
Comment (yangbodong22011) [reply src/main/java/redis/clients/jedis/DefaultJedisSocketFactory.java]: Agree, already updated.

**Review by sazzad16 — APPROVED** [review]

File: src/main/java/redis/clients/jedis/DefaultJedisSocketFactory.java [hunk 2]
```diff
@@ -75,15 +84,8 @@ private void connectToFirstSuccsefulHost(Socket socket, HostAndPort hostAndPort)
...
-      socket = new Socket();
-      socket.setReuseAddress(true);
-      socket.setKeepAlive(true); // Will monitor the TCP connection is valid
-      socket.setTcpNoDelay(true); // Socket buffer Whetherclosed, to ensure timely delivery of data
-      socket.setSoLinger(true, 0); // Control calls close () method, the underlying socket is closed immediately
-
       HostAndPort _hostAndPort = getSocketHostAndPort();
-      connectToFirstSuccsefulHost(socket, _hostAndPort);
-      socket.setSoTimeout(socketTimeout);
+      socket = connectToFirstSuccsefulHost(_hostAndPort);
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/DefaultJedisSocketFactory.java:88]:
```suggestion
      socket = connectToFirstSuccsefulHost(_hostAndPort);
      socket.setSoTimeout(socketTimeout);
```

File: src/main/java/redis/clients/jedis/DefaultJedisSocketFactory.java [hunk 3]
```diff
@@ -54,16 +54,25 @@ public DefaultJedisSocketFactory(HostAndPort hostAndPort, JedisClientConfig conf
...
     JedisConnectionException jce = new JedisConnectionException("Failed to connect to any host resolved for DNS name.");
     for (InetAddress host : hosts) {
       try {
+        Socket socket = new Socket();
+
+        socket.setReuseAddress(true);
+        socket.setKeepAlive(true); // Will monitor the TCP connection is valid
+        socket.setTcpNoDelay(true); // Socket buffer Whetherclosed, to ensure timely delivery of data
+        socket.setSoLinger(true, 0); // Control calls close () method, the underlying socket is closed immediately
+        socket.setSoTimeout(socketTimeout);
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/DefaultJedisSocketFactory.java:72]:
```suggestion
```

**Review by sazzad16 — CHANGES_REQUESTED** [review]
We used to `setSoTimeout` after successful `connect`.
https://github.com/redis/jedis/blob/4.0/src/main/java/redis/clients/jedis/DefaultJedisSocketFactory.java

**Comment by sazzad16** [comment]
@yangbodong22011

**Comment by yangbodong22011** [comment]
> @yangbodong22011

@sazzad16 Thanks for you suggestions, merged.

**Review by sazzad16 — APPROVED** [review]

**Comment by why168** [comment]
```java
org.springframework.data.redis.RedisConnectionFailureException: Cannot get Jedis connection
	at org.springframework.data.redis.connection.jedis.JedisConnectionFactory.fetchJedisConnector(JedisConnectionFactory.java:873)
	at org.springframework.data.redis.connection.jedis.JedisConnectionFactory.getConnection(JedisConnectionFactory.java:835)
	at org.springframework.data.redis.core.RedisConnectionUtils.fetchConnection(RedisConnectionUtils.java:194)
	at org.springframework.data.redis.core.RedisConnectionUtils.doGetConnection(RedisConnectionUtils.java:143)
	at org.springframework.data.redis.core.RedisConnectionUtils.getConnection(RedisConnectionUtils.java:104)
	at org.springframework.data.redis.core.RedisTemplate.execute(RedisTemplate.java:383)
	at org.springframework.data.redis.core.RedisTemplate.execute(RedisTemplate.java:363)
	at org.springframework.data.redis.core.AbstractOperations.execute(AbstractOperations.java:97)
	at org.springframework.data.redis.core.DefaultValueOperations.get(DefaultValueOperations.java:50)
	at com.lanmeishijie.jimu.swiss.service.impl.SwissServiceImpl.calculate(SwissServiceImpl.java:93)
	at com.lanmeishijie.jimu.swiss.controller.SwissController.calculate(SwissController.java:59)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:255)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:188)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:118)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:925)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:830)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1089)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:979)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:903)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:564)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:658)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:205)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149)
	at org.springframework.web.filter.ServerHttpObservationFilter.doFilterInternal(ServerHttpObservationFilter.java:109)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:167)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:482)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:115)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:673)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:344)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:391)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:896)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1744)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
	at java.base/java.lang.VirtualThread.run(VirtualThread.java:309)
Caused by: redis.clients.jedis.exceptions.JedisConnectionException: Failed to connect to any host resolved for DNS name.
	at redis.clients.jedis.DefaultJedisSocketFactory.connectToFirstSuccessfulHost(DefaultJedisSocketFactory.java:63)
	at redis.clients.jedis.DefaultJedisSocketFactory.createSocket(DefaultJedisSocketFactory.java:87)
	at redis.clients.jedis.Connection.connect(Connection.java:195)
	at redis.clients.jedis.Connection.initializeFromClientConfig(Connection.java:393)
	at redis.clients.jedis.Connection.<init>(Connection.java:68)
	at redis.clients.jedis.Jedis.<init>(Jedis.java:220)
	at redis.clients.jedis.JedisFactory.makeObject(JedisFactory.java:170)
	at org.apache.commons.pool2.impl.GenericObjectPool.create(GenericObjectPool.java:566)
	at org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:306)
	at org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:233)
	at redis.clients.jedis.util.Pool.getResource(Pool.java:38)
	at redis.clients.jedis.JedisPool.getResource(JedisPool.java:378)
	at redis.clients.jedis.JedisPool.getResource(JedisPool.java:17)
	at org.springframework.data.redis.connection.jedis.JedisConnectionFactory.fetchJedisConnector(JedisConnectionFactory.java:863)
	... 60 common frames omitted
	Suppressed: java.net.SocketTimeoutException: Connect timed out
		at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:546)
		at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:592)
		at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
		at java.base/java.net.Socket.connect(Socket.java:751)
		at redis.clients.jedis.DefaultJedisSocketFactory.connectToFirstSuccessfulHost(DefaultJedisSocketFactory.java:73)
		... 73 common frames omitted
```

**Pull request merged**